README
   Fastresolve is a package of programs to process web log files to
   get DNS information for log analysis. It sends out many queries in
   parallel, and caches results, for speed. Its DNS resolver is about
   10 times as fast as logresolve.pl from the Apache 1.3.9
   distribution, and about 50 times as fast as logresolve.c from that
   distribution. When last tested, it is 5-20 times as fast as the
   adnslogres program that comes with the adns library, and about 3
   times as fast as jdresolve, which is a Perl program that uses
   similar algorithms (https://github.com/jdrowell/jdresolve).
   
   Fastresolve is written in C++ and Perl. Versions of it have been tested
   on GNU/Linux, FreeBSD, and Solaris.
   
   Fastresolve consists of the following programs:
   
   dns-terror
          Reads log files, resolves the IP addresses that are resolvable,
          and optionally writes the results back out. Optionally it reads
          and saves the results in a DB file.
          
   reresolve
          Makes another pass trying to resolve unresolved IP addresses in
          a DB file, using dns-terror.
          
   convert-ip-db
          Converts the DB file written by dns-terror into an ASCII Analog
          DNS cache file. This is an alternative to rewriting the (huge)
          log files with the IP addresses resolved in them. On the other
          hand, Analog then has to do lookups from its big DNS cache
          file.
          
   make-report
          Runs the above programs and Analog to produce a web report.
          
   rebuild-ip-db
          Recreate the DB files from the ASCII files written by the
          convert-ip-db script, for recovering from corrupted DB files.
          
   expire-ip-db
          Removes old entries from the DB file written by dns-terror.
          
   Fastresolve requires the following:
   
   Perl.
          Any version 5 from the last few years should work.
          
   A C++ compiler.
          A recent version supporting the Standard Template Library is
          required. Clang, EGCS, GCC 2.95, or other recent GCC should
          work. (GCC 2.7.x is not recent enough; its support for
          exceptions is buggy.)
          
   The Berkeley DB library, version 2.x-5.x.
          It's probably best to get the latest (2.x, 3.x, 4.x, or 5.x) version.
          The DB library must be configured with --enable-cxx in order to
          support the C++ interface.

   The Perl BerkeleyDB module.
          
   ADNS, an asynchronous resolver library. (http://www.chiark.greenend.org.uk/~ian/adns/)
          Fastresolve comes with a patch to apply to ADNS to improve
          performance.
          
   zlib, a compression library.
          Version 1.0.6 or newer is required.
          
   Analog 3.0 or later.
          Needed by make-report.
          
   Fastresolve is distributed under the GNU General Public License.
   
   Please send bug reports, improvements, suggestions, etc. to
   fastresolve-bugs@djmnet.org.



NEWS
Changes in release 2.11:

* Support Berkeley DB 5.x.
* Fixed warnings from recent compilers.
* Support for building .rpm and .deb packages.
* Update Autoconf files.
* Update patch for adns 1.6.

Changes in release 2.10:

* Support gcc 3.2 requirement for "using namespace std;" to use STL.
* Support Sleepycat Berkeley DB 4.x.

Changes in release 2.9:

* Replaced calls to the Perl DB_File package with calls to the Perl
  BerkeleyDB package, which on some systems is easier to link with
  the same version of Berkeley DB as fastresolve.
* make-report no longer does whois queries, because getdominfo has not been
  updated to query multiple registrars' whois servers.
* The analog configuration file for make-report has been updated for analog 5.

Changes in release 2.8:

* Fixed a dns-terror memory access error on lines with nothing after
  the IP address.

Changes in release 2.7:

* dns-terror can use the Berkeley DB 3.x C++ interface as an alternative to 2.x.
* dns-terror ignores NUL bytes in an IP address field, instead of crashing,
  and passes through NUL bytes elsewhere instead of truncating the log entry.
* getdominfo retries instead of exiting if it can't connect a whois socket.

Changes in release 2.6:

* Adjust to changes in whois server responses.
* adns patch updated for 0.8.

Changes in release 2.5:

* make-report produces gzipped log files with IP addresses resolved,
  as well as the Analog report.
* convert-dom-db works around an Analog 4.0 bug that rejects some
  valid SUBDOMAIN commands.
* adns patch updated for 0.7.
* A hopefully more reliable workaround for the "adns_wait hangs" problem.
* Now available in RPM format (both source and binary).

Changes in release 2.4:

* dns-terror prints the number of DB file cache hits.
* Added dns-terror -z option to gzip the output.
* Added a workaround in dns-terror to adns hanging on the last query.
* expire-ip-db creates its temp file in the same directory as the original.
* Added expire-ip-db -m option for a more concise progress report.

Changes in release 2.3:

* Added reresolve program.
* dns-terror error handling and progress messages were improved.
* dns-terror builds cleanly on Solaris.
* getdominfo uses a simpler, more robust correct approach to parallelism.
* getdominfo adjusted for the latest whois changes.
* expire-ip-db avoids DB file corruption.
* adns patch updated for 0.6.

Changes in release 2.2:

* Added dns-terror -f option to skip fields before IP address, to
  support IIS (W3C) log files.
* dns-terror added stats to show how many of the queries were successful.
* getdominfo isn't fooled by the new whois legal notice.
* rebuild-ip-db handles the values correctly.

Changes in release 2.1:

* Added dns-terror -r option to only read positive caches from DB file.
* Simplified Analog date handling code in convert-ip-db.
* Kludged around a problem generating the make-report man page.


dns-terror/fgetln.c
/*
 * Written by David MacKenzie <djm@djmnet.org>
 * Please send comments and bug reports to <fastresolve-bugs@djmnet.org>
 *
 ******************************************************************************
 *   Copyright 1999 UUNET, an MCI WorldCom company.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 ******************************************************************************
 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>

/* Some compiler systems don't like empty modules, so ensure something's here.
 * I'm too lazy to exclude this file with configure.
 */
int dummy;

#ifndef HAVE_FGETLN
int getstr(char **lineptr, size_t *n, FILE *stream,
		      char terminator, size_t offset);

/* For BSD4.4 compatibility.  */
char *
fgetln(FILE *stream, size_t *lenp)
{
  static char *p = NULL;
  static size_t psize = 0;
  ssize_t len;

  len = getstr(&p, &psize, stream, '\n', 0);
  *lenp = len;
  return len == -1 ? NULL : p;
}
#endif


dns-terror/getline.c
/* getline.c -- Replacement for GNU C library function getline

Copyright (C) 1993, 1996 Free Software Foundation, Inc.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */

/* Written by Jan Brittenson, bson@gnu.ai.mit.edu.  */

#if HAVE_CONFIG_H
# include <config.h>
#endif

/* The `getdelim' function is only declared if the following symbol
   is defined.  */
#define _GNU_SOURCE	1
#include <stdio.h>
#include <sys/types.h>

#if defined __GNU_LIBRARY__ && HAVE_GETDELIM

int
getline (lineptr, n, stream)
     char **lineptr;
     size_t *n;
     FILE *stream;
{
  return getdelim (lineptr, n, '\n', stream);
}


#endif /* dns-terror uses getstr but not getdelim or getline */

# define NDEBUG
# include <assert.h>

#include <stdlib.h>

/* Always add at least this many bytes when extending the buffer.  */
# define MIN_CHUNK 64

/* Read up to (and including) a TERMINATOR from STREAM into *LINEPTR
   + OFFSET (and null-terminate it). *LINEPTR is a pointer returned from
   malloc (or NULL), pointing to *N characters of space.  It is realloc'd
   as necessary.  Return the number of characters read (not including the
   null terminator), or -1 on error or EOF.  */

int
getstr (lineptr, n, stream, terminator, offset)
     char **lineptr;
     size_t *n;
     FILE *stream;
     char terminator;
     size_t offset;
{
  int nchars_avail;		/* Allocated but unused chars in *LINEPTR.  */
  char *read_pos;		/* Where we're reading into *LINEPTR. */
  int ret;

  if (!lineptr || !n || !stream)
    return -1;

  if (!*lineptr)
    {
      *n = MIN_CHUNK;
      *lineptr = malloc (*n);
      if (!*lineptr)
	return -1;
    }

  nchars_avail = *n - offset;
  read_pos = *lineptr + offset;

  for (;;)
    {
      register int c = getc (stream);

      /* We always want at least one char left in the buffer, since we
	 always (unless we get an error while reading the first char)
	 NUL-terminate the line buffer.  */

      assert(*n - nchars_avail == read_pos - *lineptr);
      if (nchars_avail < 2)
	{
	  if (*n > MIN_CHUNK)
	    *n *= 2;
	  else
	    *n += MIN_CHUNK;

	  nchars_avail = *n + *lineptr - read_pos;
	  *lineptr = realloc (*lineptr, *n);
	  if (!*lineptr)
	    return -1;
	  read_pos = *n - nchars_avail + *lineptr;
	  assert(*n - nchars_avail == read_pos - *lineptr);
	}

      if (c == EOF || ferror (stream))
	{
	  /* Return partial line, if any.  */
	  if (read_pos == *lineptr)
	    return -1;
	  else
	    break;
	}

      *read_pos++ = c;
      nchars_avail--;

      if (c == terminator)
	/* Return the line.  */
	break;
    }

  /* Done - NUL terminate and return the number of chars read.  */
  *read_pos = '\0';

  ret = read_pos - (*lineptr + offset);
  return ret;
}

#if 0 /* dns-terror doesn't use these */
int
getline (lineptr, n, stream)
     char **lineptr;
     size_t *n;
     FILE *stream;
{
  return getstr (lineptr, n, stream, '\n', 0);
}

int
getdelim (lineptr, n, delimiter, stream)
     char **lineptr;
     size_t *n;
     int delimiter;
     FILE *stream;
{
  return getstr (lineptr, n, stream, delimiter, 0);
}
#endif


modules/AddrCycle.pm
# A circular list of the IP addresses configured on the local machine.
# Written by David MacKenzie <djm@djmnet.org>
# Please send comments and bug reports to <fastresolve-bugs@djmnet.org>

##############################################################################
#   Copyright 1999 UUNET, an MCI WorldCom company.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
##############################################################################

package AddrCycle;
require Exporter;
@ISA = qw(Exporter);
@EXPORT = qw(new next_addr cycle_length);

# Return one of the local machine's configured IP addresses, round-robin,
# in ASCII form.
sub next_addr
{
    my($self) = @_;
    my($addr);

    $addr = $Addrs[$self->{indx}];
    if (++$self->{indx} >= @Addrs) {
	$self->{indx} = 0;
    }
    return $addr;
}

# Return the number of IP addresses in the list.
sub cycle_length
{
    return scalar(@Addrs);
}

sub add_addr
{
    my($addr, $weedout) = @_;

    return if ($addr eq "127.0.0.1");
    return if ($weedout && $addr =~ $weedout);
    push(@Addrs, $addr);
}

sub init
{
    my($weedout) = @_;
    my($iflist);

    @Addrs = ();
    $iflist = `/sbin/ifconfig -a`;
    $iflist =~ s/\sinet\s[a-zA-Z_:]*\s*([\d.]+)\s/&add_addr($1, $weedout)/ges;
    die "AddrCycle: No inet interfaces configured" unless @Addrs;
}

# Optional arg is a regex to match local IP addresses to skip.
sub new
{
    shift;
    my $self = bless {};
    $self->{indx} = 0;
    init(@_) unless defined @Addrs;
    return $self;
}

1;


dns-terror/BoolString.h
#ifndef _BOOLSTRING_H
#define _BOOLSTRING_H

/*
 * An efficient container for a C string and a bool.
 * Callers [de]allocate memory for the string.
 * Only implements the operations that we need.
 *
 * Written by David MacKenzie <djm@djmnet.org>
 * Please send comments and bug reports to <fastresolve-bugs@djmnet.org>
 *
 ******************************************************************************
 *   Copyright 1999 UUNET, an MCI WorldCom company.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 ******************************************************************************
 */

#include <stdlib.h>
#include <string.h>
#include <iostream>

class BoolString
{
private:
  const char *str;
  bool flag;
public:
  BoolString(void) { str=0; flag=false; }
  BoolString(const char *astr, bool aflag) {
    str = astr;
    flag = aflag;
  }
  BoolString(const BoolString &other) {
    str = other.str;
    flag = other.flag;
  }
  void set_str(const char *astr) { str = astr; }
  const char *get_str(void) { return str; }
  void set_flag(bool aflag) { flag = aflag; }
  bool get_flag(void) { return flag; }

  BoolString &operator=(const BoolString &other) {
    str = other.str;
    flag = other.flag;
    return *this;
  }
  int operator<(const BoolString &other) const {
    return strcmp(str, other.str) < 0;
  }
  int operator>(const BoolString &other) const {
    return strcmp(str, other.str) > 0;
  }
  int operator!=(BoolString &other) const {
    return strcmp(str, other.str) != 0;
  }
  int operator==(BoolString &other) const {
    return strcmp(str, other.str) == 0;
  }
};

#endif /* _BOOLSTRING_H */


dns-terror/DatedStringDb.h
#ifndef _DATEDSTRINGDB_H
#define _DATEDSTRINGDB_H

/*
 * Similar to perl's hash data type, with a persistent store.
 * For storing null-terminated strings.
 *
 * Written by David MacKenzie <djm@djmnet.org>
 * Please send comments and bug reports to <fastresolve-bugs@djmnet.org>
 *
 ******************************************************************************
 *   Copyright 1999 UUNET, an MCI WorldCom company.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 ******************************************************************************
 */

#define __WANT_DB2__
#include <db_cxx.h>

class DatedStringDb
{
 public:
  DatedStringDb(const char *dbhome, const char *storename);
  int put(const char *key, const char *value, time_t *whenp = 0);
  int get(const char *key, char **valuep, time_t *whenp = 0);
  int del(const char *key);
  int cursor(Dbc **cursorp);
  int sync(void) { return db->sync(0); }
  int close(void) {
    int ret = db->close(0);
#if DB_VERSION_MAJOR >= 3
    env->close(0);
#endif
    return ret;
  };
  ~DatedStringDb();

 private:
  DbEnv *env;
  Db *db;
};

class DatedStringDbCursor
{
 public:
  DatedStringDbCursor(DatedStringDb *sdb) { sdb->cursor(&dbc); };
  int put(const char *key, const char *value, u_int32_t flags, time_t *whenp = 0);
  int get(char **keyp, char **valuep, u_int32_t flags, time_t *whenp = 0);
  int del(void) { return dbc->del(0); }
  int close(void) { return dbc->close(); };

 private:
  Dbc *dbc;
};

#endif /* _DATEDSTRINGDB_H */


dns-terror/DatedStringDb.cc
/*
 * Similar to perl's hash data type, with a persistent store.
 * For storing null-terminated strings.  We don't store the null
 * terminators in the DB file, for these reasons:
 * 1. Perl scripts that use the file would all have to remove them.
 * 2. The lengths of the strings are stored, so they're unnecessary.
 * 3. To save space.
 * The DB library tries to be smart by mallocing exactly the needed
 * number of bytes for return values.  So we realloc an extra byte,
 * which is normally a no-op because most malloc implementations
 * hand out memory in standard-sized chunks that are generally
 * larger than the size requested.  So we don't confuse the DB library,
 * we have to set the DB_DBT_MALLOC flag, and make the caller responsible
 * for freeing the result.
 *
 * The Db and Dbc classes have private constructors, so we can't
 * subclass them; the child classes' constructors would try to run their
 * parents' constructors.  So the compiler disallows it, and we get by
 * without inheritance.
 *
 * These methods return 0 on success, and nonzero (usually the value of 
 * errno) on failure.  See the DB library documentation.
 *
 * Written by David MacKenzie <djm@djmnet.org>
 * Please send comments and bug reports to <fastresolve-bugs@djmnet.org>
 *
 ******************************************************************************
 *   Copyright 1999 UUNET, an MCI WorldCom company.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 ******************************************************************************
 */

#include <sys/types.h>
#include <iostream>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <time.h>
#include "DatedStringDb.h"

using namespace std;

// DB Memory pool (cache) size in bytes.
#define MPOOL_SIZE (1024*1024*2)

// dbhome is the base if storename is relative, NULL otherwise.
// storename may be NULL to store the strings only in memory.
DatedStringDb::DatedStringDb(const char *dbhome, const char *storename)
{
#if DB_VERSION_MAJOR >= 3
  env = new DbEnv(0);
  env->set_error_stream(&cerr);
  env->set_errpfx(storename ? storename : "DatedStringDb");
  env->set_cachesize(0, MPOOL_SIZE, 0);
  env->open(dbhome, DB_CREATE|DB_INIT_MPOOL|DB_PRIVATE, 0644);
  db = new Db(env, 0);
  db->open(
#if DB_VERSION_MAJOR >= 5 || (DB_VERSION_MAJOR == 4 && DB_VERSION_MINOR > 0)
	   NULL,
#endif
	   storename, NULL, DB_BTREE, DB_CREATE, 0644);
#elif DB_VERSION_MAJOR == 2
  env = new DbEnv;
  env->set_error_stream(&cerr);
  env->set_errpfx(storename ? storename : "DatedStringDb");
  env->set_error_model(DbEnv::Exception); // Exception or ErrorReturn
  env->set_mp_size(MPOOL_SIZE);
  env->appinit(dbhome, NULL, DB_CREATE|DB_INIT_MPOOL|DB_MPOOL_PRIVATE);
  Db::open(storename, DB_BTREE, DB_CREATE, 0644, env, NULL, &db);
#else
#error Unsupported Berkeley DB version
#endif
}

// Only stores the timestamp if the pointer is nonzero.
int DatedStringDb::put(const char *key, const char *value, time_t *whenp)
{
  static char *tval = NULL;
  static size_t tlen = 0;
  size_t vlen = strlen(value);

  if (whenp) {
    if (tlen < vlen + sizeof(*whenp)) {
      tlen = vlen + sizeof(*whenp);
      if (tval)
	free(tval);
      tval = (char *) malloc(tlen);
      if (!tval)
	return ENOMEM;
    }
    memcpy(tval, whenp, sizeof(*whenp));
    memcpy(tval + sizeof(*whenp), value, vlen);
    value = tval;
    vlen += sizeof(*whenp);
  }

  Dbt dbkey((void *)key, strlen(key)),
     dbdata((void *)value, vlen);

  return db->put(NULL, &dbkey, &dbdata, 0);
}

// If valuep is NULL, just checks for the existence of the record.
// Otherwise, returns a string allocated with malloc,
// which the caller must free.
// Only assumes and returns the timestamp if the pointer is nonzero.
int DatedStringDb::get(const char *key, char **valuep, time_t *whenp)
{
  Dbt dbkey((void *)key, strlen(key)),
     dbdata;
  u_int32_t size;
  int ret;

  if (valuep)
    dbdata.set_flags(DB_DBT_MALLOC);
  ret = db->get(NULL, &dbkey, &dbdata, 0);
  if (ret != 0)
    return ret;

  if (valuep) {
    size = dbdata.get_size();
    if (whenp) {
      *valuep = (char *) dbdata.get_data();
      memcpy(whenp, *valuep, sizeof(*whenp));
      size -= sizeof(*whenp);
      memmove(*valuep, *valuep + sizeof(*whenp), size);
    } else {
      *valuep = (char *) realloc(dbdata.get_data(), size + 1);
      if (!*valuep)
	return ENOMEM;
    }
    (*valuep)[size] = '\0';
  }

  return 0;
}

int DatedStringDb::del(const char *key)
{
  Dbt dbkey((void *)key, strlen(key));

  return db->del(NULL, &dbkey, 0);
}

int DatedStringDb::cursor(Dbc **cursorp)
{
  return db->cursor(NULL, cursorp
#if DB_VERSION_MAJOR >= 3 || DB_VERSION_MINOR >= 6
		    , 0
#endif
		    );
}

DatedStringDb::~DatedStringDb()
{
  close();
  delete env;
}

//////////////////////////////////////////////////////////////////////////////

// Only stores the timestamp if the pointer is nonzero.
int DatedStringDbCursor::put(const char *key, const char *value, u_int32_t flags, time_t *whenp)
{
  static char *tval = NULL;
  static size_t tlen = 0;
  size_t vlen = strlen(value);

  if (whenp) {
    if (tlen < vlen + sizeof(*whenp)) {
      tlen = vlen + sizeof(*whenp);
      if (tval)
	free(tval);
      tval = (char *) malloc(tlen);
      if (!tval)
	return ENOMEM;
    }
    memcpy(tval, whenp, sizeof(*whenp));
    memcpy(tval + sizeof(*whenp), value, vlen);
    value = tval;
    vlen += sizeof(*whenp);
  }

  Dbt dbkey((void *)key, strlen(key)),
     dbdata((void *)value, vlen);

  return dbc->put(&dbkey, &dbdata, flags);
}

// Returns strings allocated with malloc, which the caller must free.
// Only assumes and returns the timestamp if the pointer is nonzero.
int DatedStringDbCursor::get(char **keyp, char **valuep, u_int32_t flags, time_t *whenp)
{
  Dbt dbkey, dbdata;
  u_int32_t size;
  int ret;

  dbkey.set_flags(DB_DBT_MALLOC);
  dbdata.set_flags(DB_DBT_MALLOC);
  ret = dbc->get(&dbkey, &dbdata, flags);
  if (ret != 0)
    return ret;

  size = dbkey.get_size();
  *keyp = (char *) realloc(dbkey.get_data(), size + 1);
  if (!*keyp)
    return ENOMEM;
  (*keyp)[size] = '\0';

  size = dbdata.get_size();
  if (whenp) {
    *valuep = (char *) dbdata.get_data();
    memcpy(whenp, *valuep, sizeof(*whenp));
    size -= sizeof(*whenp);
    memmove(*valuep, *valuep + sizeof(*whenp), size);
  } else {
    *valuep = (char *) realloc(dbdata.get_data(), size + 1);
    if (!*valuep)
      return ENOMEM;
  }
  (*valuep)[size] = '\0';

  return 0;
}
